File: app/src/main/java/com/yalantis/ui/activity/BaseMvpActivity.java
package com.yalantis.ui.activity;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.yalantis.contract.BaseMvpPresenter;
import com.yalantis.contract.BaseMvpView;

public abstract class BaseMvpActivity<T extends BaseMvpPresenter> extends BaseActivity implements BaseMvpView {

    protected T mPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenter = getPresenterInstance();
        mPresenter.attachView(this);
    }

    protected @NonNull abstract T getPresenterInstance();

    @Override
    public Context getContext() {
        return this;
    }

}

File: app/src/main/java/com/yalantis/base/BaseMvpActivity.kt
package com.yalantis.base

import android.os.Bundle

@Suppress("UNCHECKED_CAST", "LeakingThis")
abstract class BaseMvpActivity<in V : BaseMvpView, out T : BaseMvpPresenter<V>> : BaseActivity(), BaseMvpView {

    protected val mPresenter: T = presenterInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenter.attachView(this as V)
    }

    abstract fun presenterInstance(): T

}
